File: backend/airports_data.py
"""Lazy-loaded ICAO -> coordinates lookup, backed by data/airports.json
(generated offline by scripts/build_airports_json.py from the OurAirports
dataset - see that script's docstring)."""
import json

from . import paths

_cache = None


def _load():
    global _cache
    if _cache is None:
        try:
            with open(paths.resource_path("data", "airports.json"), "r", encoding="utf-8") as f:
                _cache = json.load(f)
        except (OSError, json.JSONDecodeError):
            _cache = {}
    return _cache


def lookup(icao_code):
    if not icao_code:
        return None
    return _load().get(icao_code.upper())

File: dotnet/src/FlightOpsHub.Core/AirportsData.cs
using System.Text.Json.Nodes;

namespace FlightOpsHub.Core;

/// <summary>
/// Port of backend/airports_data.py - lazy-loaded ICAO -> coordinates
/// lookup, backed by data/airports.json.
/// </summary>
public class AirportsData
{
    private readonly string _jsonPath;
    private JsonObject? _cache;

    public AirportsData(string jsonPath)
    {
        _jsonPath = jsonPath;
    }

    private JsonObject Load()
    {
        if (_cache != null) return _cache;
        try
        {
            var text = File.ReadAllText(_jsonPath);
            _cache = JsonNode.Parse(text) as JsonObject ?? new JsonObject();
        }
        catch
        {
            _cache = new JsonObject();
        }
        return _cache;
    }

    public JsonObject? Lookup(string? icaoCode)
    {
        if (string.IsNullOrEmpty(icaoCode)) return null;
        return Load()[icaoCode.ToUpperInvariant()] as JsonObject;
    }
}
